presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java

```diff
@@ -70,4 +70,12 @@ public BucketFunction getBucketFunction(ConnectorTransactionHandle transactionHa
     {
         throw new UnsupportedOperationException();
     }
+
+    @Override
+    public int getBucketCount(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
+    {
+        Set<Node> nodes = nodeManager.getRequiredWorkerNodes();
```

Comment:
One theoretic caveat here is the number of buckets, returned by `getBucketNodeMap` and `getBucketCount` can be different if the number of nodes changed during the two method calls. Since `nodeManager.getRequiredWorkerNodes()` will return different list.

However, this would only happen for 3 testing connector: BlackHole, TPCDS, TPCH as the "partitioning handle" is somewhat artificial in these connectors. For "real" connector partitioning handle, the number of buckets should be a constant property in the partitioning handle, thus. it's not a problem.

I think the right long term solution is to allow connector partitioning handle to return the number of buckets :). So the current implementation looks OK.

BTW: can we just throw in `getBucketCount` for BlackHole, TPCDS and TPCH connector?

Reply:
> One theoretic caveat here is the number of buckets, returned by getBucketNodeMap and getBucketCount can be different if the number of nodes changed during the two method calls. Since nodeManager.getRequiredWorkerNodes() will return different list.

Yeah, that is in theory possible. However it is also possible that `getBucketNodeMap` called multiple times may return different bucket count. But I agree, that is a little brittle. 

> BTW: can we just throw in getBucketCount for BlackHole, TPCDS and TPCH connector?

I would prefer to keep it as is, as it is sometimes useful to have TPC-DC and TPC-H connectors enabled in Presto on Spark.

Reply:
> Yeah, that is in theory possible. However it is also possible that getBucketNodeMap called multiple times may return different bucket count. But I agree, that is a little brittle.

Right. Long long ago (before grouped execution), `getBucketNodeMap` is only called once in execution. Now we already called once in `PlanFragmenter`, once in execution. 

Anyhow. :)